test/cpp/tensorexpr/tutorial.cpp

```diff
@@ -0,0 +1,366 @@
+    //     }
+    //   }
+    // }
+
+    // TODO: List all available transformations
+    // TODO: Show how statements can be constructed manually
+  }
+
+  std::cout << "*** Codegen ***" << std::endl;
+  {
+    // An ultimate goal of tensor expressions is to be provide a mechanism to
+    // execute a given computation in the fastest possible way. So far we've
+    // looked at how we could describe what computation we're interested in, but
+    // we haven't looked at how to actually execute it. So far all we've been
+    // dealing with was just symbols with no actual data associated, in this
+    // section we would look at how we can bridge that gap.
+
+    // Let's start by constructing a simple computation for us to work with:
+    Placeholder A("A", kInt, {64, 32});
+    Placeholder B("B", kInt, {64, 32});
+    Tensor* X = Compute(
+        "X",
+        {{64, "i"}, {32, "j"}},
+        [&](const VarHandle& i, const VarHandle& j) {
+          return A.load(i, j) + B.load(i, j);
+        });
+
+    // And let's lower it to a loop nest, as we did in the previous section:
+    LoopNest loopnest({X});
+    std::cout << *loopnest.root_stmt() << std::endl;
+    // Prints:
+    // {
+    //   for (int i = 0; i < 64; i++) {
+    //     for (int j = 0; j < 32; j++) {
+    //       X[i, j] = (A[i, j]) + (B[i, j]);
+    //     }
+    //   }
+
+    // Now imagine that we have two actual tensors 64x32 that we want sum
+    // together, how do we pass those tensors to the computation and how do we
+    // carry it out?
+    //
+    // Codegen object is aimed at providing exactly that functionality. Codegen
+    // is an abstract class and concrete codegens are derived from it.
+    // Currently, we have three codegens:
+    //  1) Simple Evaluator,
+    //  2) LLVM Codegen for CPU,
+    //  3) CUDA Codegen.
+    // In this example we will be using Simple Evaluator, since it's available
+    // everywhere.
+
+    // To create a codegen, we need to provide the statement - it specifies the
+    // computation we want to perform - and a list of placeholders and tensors
+    // used in the computation. The latter part is crucial since that's the only
+    // way the codegen could use to correlate symbols in the statement to actual
+    // data arrays that we will be passing when we will actually be performing
+    // the computation.
+    //
+    // Let's create a Simple IR Evaluator codegen for our computation:
+    SimpleIREvaluator ir_eval(loopnest.root_stmt(), {A, B, X});
+
+    // We are using the simplest codegen and in it almost no work is done at the
+    // construction step. Real codegens such as CUDA and LLVM perform
+    // compilation during that stage so that when we're about to run the
+    // computation everything is ready.
+
+    // Let's now create some inputs and run our computation with them:
+    std::vector<int> data_A(64 * 32, 3); // This will be the input A
```

Comment:
are strides / non-contiguity ever a thing?

Reply:
We don't have strides/non-contiguity concepts at the moment, but @asuhan had some thoughts on whether we should have them.

Reply:
@eellison We might need them, or not, depending on other decisions. Non-default strides can be used to lower `transpose` as a no-op in NNC. For example, for `x.transpose(0, 1)`, just permute the strides at positions 1 and 0. This will change address calculations for all subsequent operations on `x`, no data is moved. I'm not sure how well this will work in practice - for example, BERT calls `contiguous` right after `transpose`, so it'd defeat this trick. As of now we don't need much / anything from NNC to support this trick though, so it doesn't hurt to document it.

Reply:
@mzolotukhin if you don't have non-contiguity concepts, does it mean that if someone passes discontiguous tensor as an input `.contiguous()` would need to be called on it first?

Reply:
@ngimel, as of now, we will reject such value in the fuser pass since we're checking for complete static shapes there (and I believe/hope we check for contiguity). How we support that in a long term has not been decided yet, but if we don't want to represent strides, then - you are right - we will have to either require all inputs to be contiguous or call `.contiguous` on them in the fused kernel.

Reply:
Do you guys have tests verifying that discontiguous inputs will be rejected?

Reply:
I take back what I said about rejecting non-contiguous. We do support non-contiguous inputs, even though we don't have a concept representing strides in tensor expressions. The way it works is that we generate a valid index expression when we lower an access to a non-contiguous value:
https://github.com/pytorch/pytorch/blob/546aab66c1496845134538e78d6653fc4064ad63/torch/csrc/jit/tensorexpr/kernel.cpp#L1462-L1474

When you asked about the test I remembered that I wrote one exactly for that :)
https://github.com/pytorch/pytorch/blob/546aab66c1496845134538e78d6653fc4064ad63/test/cpp/tensorexpr/test_kernel.cpp#L102-L107